### PR TITLE
Upgrade Tekton Model to v0.47.0

### PR DESCRIPTION
**Pull request by dziemba**
## Description
- Upgrade Tekton Pipelines to v0.47.0
- Upgrade Tekton Triggers to v0.23.0
- Add the new `v1` Tekton Pipelines API Group
- Merge generator code into single package per module (to avoid duplicating boilerplate + version drifts between API groups) and remove obsolete Golang setup (e.g. vendoring)
- Rename `ArrayOrString` to `ParamValue` (has been renamed upstream)
- Fix various tests, delete obsolete ones

## Type of change

 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [x] Feature (non-breaking change which adds functionality)
 - [x] Breaking change (fix or feature that would cause existing functionality to change
 - [x] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [ ] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by rohanKanojia** [comment]
@dziemba : I think we should keep old generators v1alpha1 and v1beta1 for backward compatibility.

**Comment by dziemba** [comment]
First off: This whole PR turned out a bit larger than I anticipated. If you like me to split up some things or revert some of the (opinionated) changes, I'm happy to do that :) 

> I think we should keep old generators v1alpha1 and v1beta1 for backward compatibility.

Do you mean keeping them pinned to the older Tekton versions? I thought about that, and backwards compatibility would be super nice to have. The problem I see with that approach though is how Tekton creates new APIs. They start in alpha, then get promoted to beta and then to stable apparently. 

Screenshot 2023-03-23 at 10 27 22

Looking at the changes for v1alpha1, we are losing most of the old objects, which breaks backwards compatibility. But we also get two new objects, which we wouldn't be able to use if we're pinning v1alpha1 to an older Tekton version.

I think ideally we would need to have multiple packages for different Tekton versions, that each contain all API group models for that given version. That will quickly become a lot of code and maintenance effort though. Especially since Tekton is still on v0.x.

How do you suggest we proceed here? I'm happy to revert the changes to v1alpha1 back to how it was, but I'm a bit worried that won't be very future-proof.

**Comment by rohanKanojia** [comment]
@shawkins @manusa : Could you please provide some guidance on what should be the appropriate strategy here? Shall we keep old apiGroups to older versions or switch to the latest versions.

**Comment by dziemba** [comment]
@rohanKanojia @shawkins @manusa Any update on which direction to go in? I'm happy to change this PR as you see fit :)

**Comment by dziemba** [comment]
@vdemeester @afrittoli @bobcatfish ([Tekton topical owners for API Versioning](https://github.com/tektoncd/pipeline/blob/main/topical-ownership.md)) do you have any suggestions on what the right approach here is?

**Comment by vdemeester** [comment]
I am not sure I fully followed but tekton follow the same practices as k8s. API versions go from alpha to stable. Some alpha types can be removed as, as state in most api practice, alpha is.. breakable, from one release to the other, breaking changes on field or removal of a type can happen. 

I think a library should focus more on the api it supports than tekton version themselves. A gist from the [doc](https://github.com/tektoncd/pipeline/blob/main/api_compatibility_policy.md) 
- a `v1` type today should work the same in one year (~10/12 releases) as everyting has to be done in a backward compatible way
- a `v1beta1` type today should work the same in 9 month — after that "some" fields could be removed if properly marked as deprecated
- a `v1alpha1` type can break tomorrow

As of today (0.46), most types are either in `v1` (the main ones, like `Pipeline`, `Task`, …) and `v1beta1` (`CustomRun`, …).

File: extensions/tekton/examples/src/main/java/io/fabric8/tekton/api/examples/TaskCreate.java
Comment (rohanKanojia) [inline extensions/tekton/examples/src/main/java/io/fabric8/tekton/api/examples/TaskCreate.java:24]: I can still see v1beta1 Task in generated resources. Why do we need to delete this example?

File: extensions/tekton/examples/src/main/java/io/fabric8/tekton/api/examples/v1beta1/TaskRunCreate.java
Comment (rohanKanojia) [inline extensions/tekton/examples/src/main/java/io/fabric8/tekton/api/examples/v1beta1/TaskRunCreate.java:56]: I can still see v1beta1 TaskRun in generated resources. Why do we need to delete this example?

File: extensions/tekton/tests/src/test/java/io/fabric8/tekton/pipeline/v1beta1/TaskTest.java
Comment (rohanKanojia) [inline extensions/tekton/tests/src/test/java/io/fabric8/tekton/pipeline/v1beta1/TaskTest.java:35]: This doesn't look okay to me. Could you please create a separate test for v1 Task?

File: extensions/tekton/generator/go.mod
Comment (rohanKanojia) [inline extensions/tekton/generator/go.mod:5]: I see that [0.47.0](https://github.com/tektoncd/pipeline/tree/v0.47.0) also got released a few days ago. Shall we bump to latest version?

File: extensions/tekton/generator/cmd/generate/generate-v1beta1.go
Comment (rohanKanojia) [inline extensions/tekton/generator/cmd/generate/generate-v1beta1.go:32]: I also see a new resource `ResolutionRequest` in `resolution.tekton.dev/v1beta1` apiGroup. Do you think we should add it?

File: extensions/tekton/tests/src/test/java/io/fabric8/tekton/pipeline/v1beta1/TaskTest.java
Comment (dziemba) [reply extensions/tekton/tests/src/test/java/io/fabric8/tekton/pipeline/v1beta1/TaskTest.java:35]: Ah yes, this is still in the `v1beta1` namespace, that's no good. Should I duplicate this test for v1 and leave the v1beta1 one as-is?

File: extensions/tekton/examples/src/main/java/io/fabric8/tekton/api/examples/TaskCreate.java
Comment (dziemba) [reply extensions/tekton/examples/src/main/java/io/fabric8/tekton/api/examples/TaskCreate.java:24]: I tried to run these and they didn't work. I thought it makes sense to clean up broken examples. I can have another look at this if you want.

**Comment by dziemba** [comment]
Thanks for the review, I'll address the feedback soon

File: extensions/tekton/generator/cmd/generate/generate-v1beta1.go
Comment (dziemba) [reply extensions/tekton/generator/cmd/generate/generate-v1beta1.go:32]: I'll have a look at that

File: extensions/tekton/examples/src/main/java/io/fabric8/tekton/api/examples/v1beta1/TaskRunCreate.java
Comment (dziemba) [reply extensions/tekton/examples/src/main/java/io/fabric8/tekton/api/examples/v1beta1/TaskRunCreate.java:56]: I restored this file

File: extensions/tekton/examples/src/main/java/io/fabric8/tekton/api/examples/TaskCreate.java
Comment (dziemba) [reply extensions/tekton/examples/src/main/java/io/fabric8/tekton/api/examples/TaskCreate.java]: I restored this file

File: extensions/tekton/tests/src/test/java/io/fabric8/tekton/pipeline/v1beta1/TaskTest.java
Comment (dziemba) [reply extensions/tekton/tests/src/test/java/io/fabric8/tekton/pipeline/v1beta1/TaskTest.java]: I restored this as-is and make a new test for the `v1` one

**Comment by dziemba** [comment]
@rohanKanojia I addressed all feedback now, please have another look :)

**Comment by manusa** [comment]
Hi @dziemba,

Sorry for the late feedback. This morning we noticed that some of the structs are missing.

More specifically, the Interceptor in the triggers module (https://pkg.go.dev/github.com/tektoncd/triggers@v0.23.1/pkg/apis/triggers/v1alpha1#Interceptor).

Could you please add that one too?

**Comment by dziemba** [comment]
@manusa Sure, I've now added the `Interceptor` types.

**Comment by manusa** [comment]
> @manusa Sure, I've now added the `Interceptor` types.
> 
> Update: I forgot to add the new type to the API DSL, I just pushed a new commit that includes these changes as well.

Hi @dziemba, sorry for the late updates. We've been trying to release a stable 6.6 version.

If you covered everything we mentioned, then it should be good to go. Don't worry about changelog conflicts or similar. If you checked the `Allow edits and access to secrets by maintainers` I should be able to take care of that.

**Comment by rohanKanojia** [comment]
This is also affected by https://github.com/fabric8io/kubernetes-client/issues/5113 . We should refactor trigger generators to remove the split package warning.

**Review by rohanKanojia — APPROVED** [review]

**Comment by metacosm** [comment]
Should we update triggers to 0.24 that got released last week?

**Comment by manusa** [comment]
CHANGELOG was not correctly updated, I'll fix after merging to avoid re-running all pipelines.

Improvement entries need to be removed

**Review by manusa — APPROVED** [review]

**Comment by manusa** [comment]
LGTM, thx @dziemba!

**Pull request merged**
